File: 0875-koko-eating-bananas/0875-koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        ans=[]
        st=1
        end=max(piles)
        while(st<=end):
            mid=(st+end)//2
            speed=mid
            t=0
            for i in piles:
                t+=ceil(i/speed)
            if t<=h:
                ans.append(speed)
                end=mid-1
            else:
                st=mid+1
        return min(ans)

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        // Arrays.sort(piles);
        // if(piles.length==1) return piles[0];
        int start = 1;
        int end =  Arrays.stream(piles).max().getAsInt();
        int ans = Integer.MAX_VALUE;
        while(start<=end){
            int speed = start+(end-start)/2;
            long time = 0;
            for(int pile: piles){
                time+=(pile+speed-1L)/speed;
            }
            if(time <= h){
                ans = Math.min(ans,speed);
                end = speed-1;
            }
            else start = speed+1;

            
        }
        return ans;
    }
}